routemaster/state_machine/utils.py

```diff
@@ -0,0 +1,188 @@
+"""Utilities for state machine execution."""
+
+import datetime
+from typing import Any, Dict, List, Tuple
+
+import dateutil.tz
+from sqlalchemy import and_, func, select
+
+from routemaster.db import labels, history
+from routemaster.app import App
+from routemaster.feeds import feeds_for_state_machine
+from routemaster.config import (
+    Gate,
+    State,
+    Action,
+    StateMachine,
+    ContextNextStates,
+)
+from routemaster.context import Context
+from routemaster.state_machine.types import LabelRef, Metadata
+from routemaster.state_machine.exceptions import (
+    UnknownLabel,
+    UnknownStateMachine,
+)
+
+
+def get_state_machine(app: App, label: LabelRef) -> StateMachine:
+    """Finds the state machine instance by name in the app config."""
+    try:
+        return app.config.state_machines[label.state_machine]
+    except KeyError as k:
+        raise UnknownStateMachine(label.state_machine)
+
+
+def start_state_machine(app: App, label: LabelRef, conn) -> None:
+    """Create the first history entry for a label in a state machine."""
+    state_machine = get_state_machine(app, label)
+    conn.execute(history.insert().values(
+        label_name=label.name,
+        label_state_machine=label.state_machine,
+        old_state=None,
+        new_state=state_machine.states[0].name,
+    ))
+
+
+def choose_next_state(
+    state_machine: StateMachine,
+    current_state: State,
+    context: Context,
+) -> State:
+    """Assuming a transition is valid, choose a next state."""
+    next_state_name = current_state.next_states.next_state_for_label(context)
+    return state_machine.get_state(next_state_name)
+
+
+def get_label_metadata(
+    label: LabelRef,
+    state_machine: StateMachine,
+    conn,
+) -> Tuple[Dict[str, Any], bool]:
```

Comment:
@danpalmer I'm not sure this is correct -- what happens if the label isn't found? (I'm expecting that `fetchone` returns `None` in such cases)

Reply:
Recorded as https://app.asana.com/0/465891412411656/517526618377258/f